codice/ddf
# [2.13.x] DDF-4042 Rewrite location input to support manual editing of all shapes

#### What does this PR do?
Completely overhauls the existing geometry input in Intrigue (the one used in the inspector and upload editor) so it supports the same shapes and most coordinate systems as the search builder location input. In this initial implementation, the input does not support UTM, but it does add support for DMS coordinates. Furthermore, it adds support for Point geometries, makes all shapes manually editable without the use of the map, and provides validation tips to help users enter valid geometries.

#### Who is reviewing it? 
@vinamartin @djblue
#### Select relevant component teams: 
@codice/ui
#### Ask 2 committers to review/merge the PR and tag them here.

@andrewkfiedler
@brendan-hofmann
#### How should this be tested?
Configure the upload editor to show the location attribute:
* In the Admin UI, navigate to Search UI > Configuration > Catalog UI Search
* Add an entry for `location` to the attribute configuration
* Also add an entry for `location` to the required attributes list

In Intrigue, verify that:
* The new geometry input is shown in the inspector view and the upload editor, but the search builder location input is not changed.
* The inspector displays a WKT string when not in edit mode.
* Validation works correctly. If invalid coordinates are entered, a warning should appear, and no warnings should appear for valid input.
* Input geometries are converted to WKT and assigned to the metacard.
* All shape and coordinate systems function as expected.
* If the input is blank and an upload is attempted, a warning shows to indicate the attribute is required, and it disappears as soon as the user provides input.

#### Any background context you want to provide?
The metacard editor and upload editor currently only provide WKT editing support, as that's the format used to store all geometries on the metacard. But most user's aren't familiar with WKT (ironically). The goal of this PR was to allow users to edit geometries using the standard coordinate systems that they understand. Since the editors don't have access to the map, a primary goal was to provide first-class support for manual editing.  

#### What are the relevant tickets?
[DDF-4042](https://codice.atlassian.net/browse/DDF-4042)
#### Screenshots
WKT editor (same as before)
<img width="366" alt="wkt_valid" src="https://user-images.githubusercontent.com/3878142/43914556-10f43780-9bbd-11e8-91ce-a29d872fb9e9.png">
List editor for lines and polygons
<img width="483" alt="screen shot 2018-08-17 at 5 56 06 pm" src="https://user-images.githubusercontent.com/3878142/44294098-e64cbf80-a246-11e8-978b-aaea1c4bc212.png">
DMS coordinate support
<img width="366" alt="usng_valid" src="https://user-images.githubusercontent.com/3878142/43914593-240d064e-9bbd-11e8-9adc-3cd7e0344323.png">
USNG editor
<img width="363" alt="dms_valid" src="https://user-images.githubusercontent.com/3878142/43914586-20482b06-9bbd-11e8-9c2d-836efc6a14db.png">
Inspector view and sample validation warning
<img width="1017" alt="screen shot 2018-08-09 at 10 19 39 am" src="https://user-images.githubusercontent.com/3878142/43914871-e033bcb4-9bbd-11e8-883a-1b8dfe0f6ea2.png">



#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
